packages/library-base/src/commonMain/kotlin/io/realm/internal/schema/CachedClassKeyMap.kt

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright 2021 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.realm.internal.schema
+
+import io.realm.internal.interop.ClassKey
+import io.realm.internal.interop.NativePointer
+import io.realm.internal.interop.PropertyInfo
+import io.realm.internal.interop.PropertyKey
+import io.realm.internal.interop.RealmInterop
+
+/**
+ * Schema metadata providing access to class metadata for the schema.
+ */
+interface SchemaMetadata {
+    operator fun get(className: String): ClassMetadata?
+    fun getOrThrow(className: String): ClassMetadata = get(className)
+        ?: throw IllegalArgumentException("Object of type '$className not found")
+}
+
+/**
+ * Class metadata providing access class and property keys.
+ */
+interface ClassMetadata {
+    val className: String
+    operator fun get(propertyName: String): PropertyKey?
+    fun getOrThrow(propertyName: String): PropertyKey = get(propertyName)
+        ?: throw IllegalArgumentException("Object of type '$className doesn't have a property named '$propertyName'")
+}
+
+// FIXME We are actually not lazy loading these right now
```

Comment:
Not 100% this is actually something we do want to fix...running the logic on the notifier thread will keep the Main thread faster at the close of slightly slower notifications when the schema actually does change.

Reply:
Rephrased into a `TODO OPTIMIZE`, so that we revisit this as part of optimizing when we have an appropriate benchmarking mechanism.